### PR TITLE
Bug xtime

### LINKED ISSUE #411: bug: Inconsistent output for the value of xtime variable when using GNU Fortran
:bug:

### Describe the bug

Executions of ./filter with MPAS_ATM that should be bitwise identical (i.e. have the exact same data in their output files) are not when compiling with GNU Fortran, as they display differences in their output NetCDF files for the variable xtime. This is due to the difference in the assigned length of the string xtime (64 characters) versus the actual length of the date string, which is much shorter than 64. When compiling with Intel, the empty characters at the end of the string are truncated. However, when compiling and running with GNU Fortran, the empty characters at the end of the string are instead filled with random junk from memory, which is then written to the output NetCDF files. 

Intel:
Screen Shot 2022-10-10 at 1 56 31 PM

GNU Fortran:
Screen Shot 2022-10-10 at 1 58 27 PM

1. List the steps someone needs to take to reproduce the bug. 

### Setup:
There is a small and basic MPAS_ATM test case that can be found on my Cheyenne. Running this and inspecting the output files will allow you to see the value of the xtime variable.

To use this test case, you will need to copy the /xtime_gfort_bug_tester directory into your DART repo on scratch.

Scratch is one of the available file spaces on Cheyenne, and it needs to be used when copying whole directories or large amounts of data onto your Cheyenne as opposed to your home space (/glade/u/home) because the home space has a much smaller storage quota and can not hold this directory. Here is some more info on the Cheyenne file spaces if you want to learn more: https://arc.ucar.edu/knowledge_base/68878466 

If you do not already have a DART repo on scratch, go ahead and either create a new fork or use your current fork and clone into the directory /glade/scratch/YOURUSERNAME/ 

Copy the /xtime_gfort_bug_tester directory into your fork of DART repo on Cheyenne. First, navigate to the mpas_atm folder in your DART repo on scratch.

`cd glade/scratch/YOURUSERNAME/DART/models/mpas_atm/`

then

`cp -r /glade/scratch/masmith/DART/models/mpas_atm/xtime_gfort_bug_tester/ .`

`cd xtime_gfort_bug_tester`

### Running the test case:
Make sure you are using the latest version of the GNU Fortran compiler. To do this, you must make sure that you are using the correct build_template and that you have the correct compiler module loaded on Cheyenne. 

Navigate to the /build_templates directory and copy the mkmf.template.gfortran template into your working template (mkmf.template)

`cd glade/scratch/YOURUSERNAME/DART/build_templates`

then

`cp mkmf.template.gfortran mkmf.template`

Now navigate to the work directory for mpas_atm 

`cd glade/scratch/YOURUSERNAME/DART/models/mpas_atm/work`

The work directory is where we compile/build the DART programs. 

`module list` - this command will show you the modules that are currently being used
`module av` - this command will show you all available modules

`module load gnu/12.1.0`
`./quickbuild clean`
`./quickbuild filter`

Now that you have built the filter program, navigate back to your test directory.

`cd /glade/scratch/YOURUSERNAME/DART/models/mpas_atm/xtime_gfort_bug_tester/ `

And you can no copy the filter program you just built in the work dir into the test dir.

`cp ../work/filter .`

NOTE: you will need to repeat this copy each time you compile/build filter, or alternatively you can create a symbolic link to filter that will automatically update whenever the filter program in the work directory is recompiled and built. 

`ln -s ../work/filter ./filter`

You will need to remove the old output files from the previous run of filter. Sometimes, if you do not delete them before running ./filter again, filter will not overwrite the output files and the data from the previous run will stay.

`rm obs_seq.final output_mean.nc output_priorinf_mean.nc output_priorinf_sd.nc output_sd.nc out_restart.nc_1 out_restart.nc_2 out_restart.nc_3 out_restart.nc_4 out_restart.nc_5 dart_log.out`

`qsub -I -l select=2:ncpus=36:mpiprocs=36 -l walltime=02:00:00 -q regular -A P86850054`

This will start an interactive job on Cheyenne using the specified resources, allowing you to run filter. During interactive jobs, the output of filter will be printed to the terminal as the program runs. 

***While on the interactive job:***
`module load gnu/12.1.0`
`mpirun ./filter`

Seeing the following lines at the end shows that the program has successfully finished its run.
```
 Finished ... at YYYY MM DD HH MM SS = 
                 2022 10 10 13 14  8
 Program Filter
```

`exit` - this command finishes the interactive job

### Inspecting the output files: 
`ncdump output_mean.nc > output_mean1`
`less output_mean1`

ESC, Shift+G will take you to the end of the file, where the xtime variable is written 

NOTE: you can inspect any output NetCDF file you’d like (output_mean.nc, output_priorinf_mean.nc, output_priorinf_sd.nc,  output_sd.nc, out_restart.nc_1, out_restart.nc_2, etc.) and will see the output of the variable xtime at the file; I just tend to use output_mean

2. What was the expected outcome?
The output files are identical across runs that are otherwise bitwise identical (ex: two MPAS_ATM runs in which no changes were made to the code or input values).

3. What actually happened?  
Values for the variables xtime written into output NetCDF files will vary for every execution of filter with MPAS_ATM.

### Making changes to the code and seeing their impact:
After making edits to the code, repeat the processes above for running the test case and inspecting the output files to see if there are any changes to the value of xtime. 

NOTE: As stated earlier in the issue, output files might not be overwritten if they already exist, so make sure to remove them before each run of filter. The results of the ncdump command will not need to be deleted, but just make sure you don’t accidentally overwrite them by dumping into the same file each time. After making edits to the code and running filter successfully, change the name of the output file for ncdump to output_mean2 instead of output_mean1, or use names that will help you remember which run of filter they correspond to (maybe output_mean_unedited_code or something like that).

`ncdump output_mean.nc > output_mean2`
`less output_mean2`

### Error Message  
No relevant error messages. 

### Which model(s) are you working with?
MPAS_ATM

### Version of DART      
The latest, v10.5.1

### Have you modified the DART code?   
No

### Build information
Please describe:  
 1. The machine you are running on (e.g. windows laptop, NCAR supercomputer Cheyenne).   
Cheyenne
 2. The compiler you are using (e.g. gnu, intel).  
GNU

### PR DESCRIPTION
**Pull request by achauphan**
## Description:

Fixes #411 

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Documentation changes needed?

- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.

### Tests
Please describe any tests you ran to verify your changes.

## Checklist for merging

- [ ] Updated changelog entry
- [ ] Documentation updated
- [ ] Update conf.py

## Checklist for release
- [ ] Merge into main
- [ ] Create release from the main branch with appropriate tag
- [ ] Delete feature-branch

## Testing Datasets

- [x] Dataset needed for testing available upon request
- [ ] Dataset download instructions included
- [ ] No dataset needed

### CODE REVIEW

**Comment by nancycollins** [comment]
hi anderson,

you don't want to add the log files to the repo, even though they're interesting.  you also can't alter the documentation in the readme.rst because that's showing what the mpss init files look like and it's going to be 64 when someone dumps the file. 
 the mpas_init file comes from the scientist running mpas, so you can't change the string length in it.  that's under the model's control. 

if that dart write time routine is creating a variable from scratch, i think it has to stay length 64 because mpas may be expecting that to be the case when it reads in the file.   i think you may need to set the string time with the get_time() routine and then set the 20th character to null (using char(0) ) to terminate the string.  in fortran the valid parts of the string have a known length as well as a max length, but in C the string is terminated by a null character.  if you have 19 good characters and a 20th null char that may fix it.  or you may need to initialize the entire string to char(0) first and then call the get_time() routine.  see if that works?

**Comment by achauphan** [comment]
Hi Nancy, thanks for pointing out those log files I had included. 
Your recommended approach to fixing this issue does make more sense to avoid conflicting expectations in the defined lengths for until the timestring is populated. I'll give it a go.

**Comment by nancycollins** [comment]
helen pointed out to me that using the fortran "trim()" intrinsic on the string being passed to nc_put_variable on line 3081 might be all it needs.  instead of passing **timestring** as the third arg, try passing **trim(timestring)** instead and see if the fortran routines take care of it themselves.

**Review by hkershaw-brown — CHANGES_REQUESTED** [review]
Fix looks good. 

The file `models/mpas_atm/dart_log_fail.out` needs removing from this pull request.

File: models/mpas_atm/model_mod.f90 [hunk 1]
```diff
@@ -3051,7 +3051,7 @@ subroutine write_model_time_restart(ncid, dart_time)
 type(time_type),     intent(in) :: dart_time
 
 integer :: year, month, day, hour, minute, second
-character(len=64) :: timestring
+character(len=64) :: timestring = ''
```
Comment (hkershaw-brown) [inline models/mpas_atm/model_mod.f90:3054]: Note to check before we merge this. Initializing in the declaration is an implicit save in Fortran, so `timestring` is only initialed on the first call to `write_model_time_restart`. We may want to change the initialization to be separate:

```suggestion
character(len=64) :: timestring
timestring = ''
```

**Review by hkershaw-brown — APPROVED** [review]
Great work on this bug fix Anderson. 

We'll bundle release this with #418

**Pull request merged**